observatory-platform/observatory/platform/bigquery.py

```diff
@@ -294,6 +298,71 @@ def bq_find_schema(
     return None
 
 
+def bq_generate_schema_from_data(data_path: str) -> List[dict]:
+    """
+    Generates a Biguqery style schema based on a *.jsonl.gz data file.
+
+    :param expected_schema_path: Path to the expected schema of the data.
+    :param data_path: Path to the *.jsonl.gz data file.
+    :param check_data_types: Optional, set to True to check the data type for expected vs generated.
+    :return: True if the expected and generated schemas match.
+    """
+
+    with CliRunner().isolated_filesystem() as temp_file_path:
+        generator = SchemaGenerator()
+        random_id = observatory_environment.random_id()
+        generated_schema_path = os.path.join(temp_file_path, f"generated_schema_{random_id}.json")
+        with gzip.open(data_path, "rb") as f_in, open(generated_schema_path, "w") as f_out:
+            generator.run(input_file=f_in, output_file=f_out, schema_map=None)
+        with open(generated_schema_path, "r") as f_in:
+            generated = json.load(f_in)
```

Comment:
Saving the schema to JSON with generator.run can be done in memory without using a temporary file. I think by using io.StringIO() in place of the temporary file.